File: js/packages/core/src/api-integration.ts
import { IForeventsAPIError, IForeventsQuotaExceededError, classifyResponse } from "./errors";
import { Integration, type IntegrationHooks } from "./integration";
import { MemoryStorage, type Storage } from "./storage";
import type { EventType, IdentifyEvent, Logger, PageEvent, Properties, TrackEvent } from "./types";

/**
 * Configuration of the first-party API integration.
 *
 * Only `projectKey` is required. It is a public write key: it grants event
 * ingestion and nothing else, so shipping it in a browser bundle or an app
 * binary is safe. There is deliberately no project secret here.
 */
export interface IForeventsAPIConfig {
  /** Public project write key (required). */
  projectKey: string;
  /** API origin. Default `https://api.iforevents.com`. */
  baseUrl?: string;
  /** Events to accumulate before sending (1..500). 1 disables batching. Default 10. */
  batchSize?: number;
  /** Milliseconds a partial batch waits before it is sent. Default 5000. */
  flushInterval?: number;
  /** Per-request timeout in milliseconds. Default 10000. */
  timeout?: number;
  /** Retries for transient failures (network, 5xx, short rate limits). Default 3. */
  maxRetries?: number;
  /** Base delay between retries in milliseconds, multiplied by the attempt number. Default 1000. */
  retryDelay?: number;
  /** Put events that failed with a transient error back in the queue. Default true. */
  requeueFailedEvents?: boolean;
  /** Log requests and failures. Default false. */
  debug?: boolean;
  /** Reject `identify`/`flush` instead of only logging when a request fails. Default false. */
  throwOnError?: boolean;
  /** Called once when the api starts refusing events with `quota_exceeded`, and again after a later success followed by a new refusal. */
  onQuotaExceeded?: (error: IForeventsQuotaExceededError) => void;
  /** Called for every failed request after retries are exhausted. */
  onError?: (error: IForeventsAPIError) => void;
  /** Where the user uuid and (optionally) the pending queue live. Default in-memory. */
  storage?: Storage;
  /** Persist the pending queue in `storage` so unsent events survive a restart. Default false. */
  persistQueue?: boolean;
  /** Most events kept in the persisted queue. Default 1000. */
  maxQueueSize?: number;
  /** Custom fetch, for tests or runtimes that polyfill it. Default `globalThis.fetch`. */
  fetch?: typeof fetch;
  /** Value of the `User-Agent` header where the runtime allows setting it (servers). */
  userAgent?: string;
  logger?: Logger;
}

/** One event as queued and sent to `/v1/events/batch`. */
export interface QueuedEvent {
  name: string;
  type: EventType;
  properties: Properties;
  created_at: string;
}

export interface QueueStatus {
  queuedEvents: number;
  batchSize: number;
  isInitialized: boolean;
  isIdentified: boolean;
  userId: string | null;
}

const USER_KEY = "iforevents_user_id";
const IDENTIFIED_KEY = "iforevents_user_identified";
const QUEUE_KEY = "iforevents_queue";
const MAX_BATCH = 500;
const DEFAULT_BASE_URL = "https://api.iforevents.com";

type Resolved = Required<Pick<IForeventsAPIConfig, "baseUrl" | "batchSize" | "flushInterval" | "timeout" | "maxRetries" | "retryDelay" | "requeueFailedEvents" | "debug" | "throwOnError" | "persistQueue" | "maxQueueSize">>;

/**
 * Talks to the IForevents ingest API: identify, single or batched track,
 * page views, persisted user uuid, retries with `Retry-After`, and typed
 * errors. Mirrors `IForeventsAPIIntegration` of the Flutter package.
 */
export class IForeventsAPIIntegration extends Integration {
  readonly config: IForeventsAPIConfig & Resolved;
  private readonly storage: Storage;
  private readonly fetchImpl: typeof fetch;
  private readonly logger: Logger | null;

  private queue: QueuedEvent[] = [];
  private timer: ReturnType<typeof setTimeout> | null = null;
  private inflight: Promise<void> = Promise.resolve();
  private userId: string | null = null;
  private initialized = false;
  private identified = false;
  private quotaExceeded = false;

  constructor(config: IForeventsAPIConfig, hooks: IntegrationHooks = {}) {
    super(hooks, "IForeventsAPIIntegration");
    if (!config || typeof config.projectKey !== "string" || config.projectKey.trim() === "") {
      throw new Error("[iforevents] IForeventsAPIIntegration needs a projectKey");
    }
    if ("projectSecret" in (config as object)) {
      throw new Error("[iforevents] the project secret never belongs in an SDK; pass the project key only");
    }
    const batchSize = Math.max(1, Math.min(MAX_BATCH, Math.floor(config.batchSize ?? 10)));
    this.config = {
      ...config,
      baseUrl: (config.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, ""),
      batchSize,
      flushInterval: config.flushInterval ?? 5000,
      timeout: config.timeout ?? 10000,
      maxRetries: config.maxRetries ?? 3,
      retryDelay: config.retryDelay ?? 1000,
      requeueFailedEvents: config.requeueFailedEvents ?? true,
      debug: config.debug ?? false,
      throwOnError: config.throwOnError ?? false,
      persistQueue: config.persistQueue ?? false,
      maxQueueSize: config.maxQueueSize ?? 1000,
    };
    this.storage = config.storage ?? new MemoryStorage();
    const f = config.fetch ?? (globalThis as { fetch?: typeof fetch }).fetch;
    if (typeof f !== "function") throw new Error("[iforevents] no fetch available; pass config.fetch");
    // window.fetch throws "Illegal invocation" unless called on the window.
    this.fetchImpl = f.bind(globalThis);
    this.logger = this.config.debug ? (config.logger ?? console) : null;
  }

  // --- state -----------------------------------------------------------------

  get isInitialized(): boolean {
    return this.initialized;
  }
  get isIdentified(): boolean {
    return this.identified;
  }
  /**
   * The id every request carries in `X-User-Id`: a generated `anon_...` id
   * kept per visitor, or the customId of the last identify.
   */
  get currentUserId(): string | null {
    return this.userId;
  }
  get queuedEventsCount(): number {
    return this.queue.length;
  }
  /** True after a `quota_exceeded` answer until the next accepted request. */
  get isQuotaExceeded(): boolean {
    return this.quotaExceeded;
  }
  getQueueStatus(): QueueStatus {
    return { queuedEvents: this.queue.length, batchSize: this.config.batchSize, isInitialized: this.initialized, isIdentified: this.identified, userId: this.userId };
  }

  // --- Integration ------------------------------------------------------------

  override async init(): Promise<void> {
    await super.init();
    const stored = await this.storage.get(USER_KEY);
    if (stored) {
      this.userId = stored;
      this.identified = (await this.storage.get(IDENTIFIED_KEY)) === "true";
    } else {
      // A fresh visitor: attribute everything to an anonymous id we own, so the
      // api never has to fingerprint the address (which merges users behind a NAT).
      await this.setUser(anonymousId(), false);
    }
    if (this.config.persistQueue) {
      const raw = await this.storage.get(QUEUE_KEY);
      if (raw) {
        try {
          const events = JSON.parse(raw) as QueuedEvent[];
          if (Array.isArray(events) && events.length > 0) {
            this.queue = [...events, ...this.queue].slice(-this.config.maxQueueSize);
            this.schedule();
          }
        } catch {
          await this.storage.remove(QUEUE_KEY);
        }
      }
    }
    this.initialized = true;
    this.logger?.debug("[iforevents] api integration ready", { baseUrl: this.config.baseUrl, batchSize: this.config.batchSize });
  }

  override async identify(event: IdentifyEvent): Promise<void> {
    await super.identify(event);
    const { email, name, phone_number, ...properties } = event.traits;
    const body: Record<string, unknown> = { custom_id: event.customId, properties };
    if (typeof email === "string" && email) body.email = email;
    if (typeof name === "string" && name) body.name = name;
    if (typeof phone_number === "string" && phone_number) body.phone_number = phone_number;
    // Attribute from now on, even if the profile request itself fails: the
    // api creates the profile on the first event it sees for this id.
    await this.setUser(event.customId, true);
    try {
      await this.request("/v1/events/identify", body);
    } catch (error) {
      this.report(error);
      if (this.config.throwOnError) throw error;
    }
  }

  override async track(event: TrackEvent): Promise<void> {
    await super.track(event);
    const queued: QueuedEvent = { name: event.name, type: event.type, properties: event.properties, created_at: event.timestamp.toISOString() };
    if (this.config.batchSize <= 1) {
      try {
        await this.request("/v1/events/track", { event_name: queued.name, event_type: queued.type, properties: queued.properties });
      } catch (error) {
        this.report(error);
        if (this.config.throwOnError) throw error;
      }
      return;
    }
    this.queue.push(queued);
    if (this.queue.length > this.config.maxQueueSize) this.queue.splice(0, this.queue.length - this.config.maxQueueSize);
    await this.persist();
    if (this.queue.length >= this.config.batchSize) void this.flush();
    else this.schedule();
  }

  override async page(event: PageEvent): Promise<void> {
    await super.page(event);
    const properties: Properties = { ...event.properties };
    if (event.navigationType !== undefined) properties.navigation_type = event.navigationType;
    if (event.toRoute !== undefined) properties.to_route = event.toRoute;
    if (event.previousRoute !== undefined) properties.previous_route = event.previousRoute;
    await this.track({ name: event.name || "page_view", type: "page_view", properties, timestamp: event.timestamp });
  }

  override async reset(): Promise<void> {
    await super.reset();
    await this.flush();
    // Forget the person; the next events belong to a fresh anonymous id.
    await this.setUser(anonymousId(), false);
  }

  /** Sends the whole queue now, 500 events per request. Resolves when done. */
  override async flush(options: { keepalive?: boolean } = {}): Promise<void> {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    // A previous rejected drain (throwOnError) must not poison later flushes.
    const run = this.inflight.catch(() => undefined).then(() => this.drain(options.keepalive ?? false));
    this.inflight = run.catch(() => undefined);
    await run;
  }

  override async shutdown(): Promise<void> {
    await this.flush();
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  }

  // --- internals ----------------------------------------------------------------

  private async drain(keepalive: boolean): Promise<void> {
    while (this.queue.length > 0) {
      const events = this.queue.splice(0, MAX_BATCH);
      try {
        await this.request("/v1/events/batch", { events }, keepalive);
        await this.persist();
      } catch (error) {
        const apiError = error instanceof IForeventsAPIError ? error : new IForeventsAPIError(String(error), { cause: error });
        if (apiError.retryable && this.config.requeueFailedEvents) {
          // Transient: keep these events at the front for the next flush.
          this.queue.unshift(...events);
          this.schedule();
        } else if (!apiError.retryable) {
          // A refused key or an exhausted quota would fail the same way forever: drop everything.
          this.queue = [];
        }
        await this.persist();
        this.report(apiError);
        if (this.config.throwOnError) throw apiError;
        return;
      }
    }
  }

  private schedule(): void {
    if (this.timer || this.queue.length === 0) return;
    this.timer = setTimeout(() => {
      this.timer = null;
      void this.flush();
    }, this.config.flushInterval);
    // Never keep a server process alive just for a pending batch.
    (this.timer as { unref?: () => void }).unref?.();
  }

  private async setUser(id: string, identified: boolean): Promise<void> {
    this.userId = id;
    this.identified = identified;
    await this.storage.set(USER_KEY, id);
    await this.storage.set(IDENTIFIED_KEY, identified ? "true" : "false");
  }

  private async persist(): Promise<void> {
    if (!this.config.persistQueue) return;
    if (this.queue.length === 0) await this.storage.remove(QUEUE_KEY);
    else await this.storage.set(QUEUE_KEY, JSON.stringify(this.queue));
  }

  private report(error: unknown): void {
    this.logger?.warn("[iforevents] request failed", error);
    if (error instanceof IForeventsAPIError) this.config.onError?.(error);
  }

  private noteOutcome(error: IForeventsAPIError | null): void {
    if (error instanceof IForeventsQuotaExceededError) {
      if (!this.quotaExceeded) {
        this.quotaExceeded = true;
        this.config.onQuotaExceeded?.(error);
      }
    } else if (error === null) {
      this.quotaExceeded = false;
    }
  }

  /** POSTs JSON with retries; resolves with the decoded body (or null). */
  private async request(path: string, body: unknown, keepalive = false): Promise<unknown> {
    const headers: Record<string, string> = { "Content-Type": "application/json", "X-Project-Key": this.config.projectKey };
    if (this.userId) headers["X-User-Id"] = this.userId;
    // Only headers on the api's CORS allowlist (Content-Type, X-Project-Key,
    // X-User-Id) may be sent from browsers; the sdk name and version
    // travel inside the context properties instead of a custom header.
    if (this.config.userAgent) headers["User-Agent"] = this.config.userAgent;
    const payload = JSON.stringify(body);
    const url = `${this.config.baseUrl}${path}`;

    for (let attempt = 0; ; attempt++) {
      let error: IForeventsAPIError;
      try {
        const controller = typeof AbortController === "function" ? new AbortController() : null;
        const timeoutId = controller ? setTimeout(() => controller.abort(), this.config.timeout) : null;
        let res: Response;
        try {
          res = await this.fetchImpl(url, { method: "POST", headers, body: payload, keepalive, signal: controller?.signal ?? null });
        } finally {
          if (timeoutId) clearTimeout(timeoutId);
        }
        const text = await res.text();
        const json = parseJSON(text);
        this.logger?.debug(`[iforevents] POST ${path} -> ${res.status}`);
        if (res.ok) {
          this.noteOutcome(null);
          return json;
        }
        error = classifyResponse(res.status, json, res.headers.get("retry-after"));
      } catch (cause) {
        error = new IForeventsAPIError(cause instanceof Error ? cause.message : String(cause), { cause });
      }
      if (!error.retryable || attempt >= this.config.maxRetries) {
        this.noteOutcome(error);
        throw error;
      }
      const retryAfter = "retryAfterMs" in error ? (error as { retryAfterMs?: number }).retryAfterMs : undefined;
      const delay = retryAfter && retryAfter > 0 ? retryAfter : this.config.retryDelay * (attempt + 1);
      this.logger?.debug(`[iforevents] retrying ${path} in ${delay}ms (${attempt + 1}/${this.config.maxRetries})`);
      await sleep(delay);
    }
  }
}

/** A fresh anonymous id, unrelated to anything the server derives: anon_<uuid4 without dashes>. */
export function anonymousId(): string {
  const bytes = new Uint8Array(16);
  const c = (globalThis as { crypto?: { getRandomValues?: (a: Uint8Array) => Uint8Array } }).crypto;
  if (c?.getRandomValues) c.getRandomValues(bytes);
  else for (let i = 0; i < 16; i++) bytes[i] = Math.floor(Math.random() * 256);
  bytes[6] = (bytes[6]! & 0x0f) | 0x40;
  bytes[8] = (bytes[8]! & 0x3f) | 0x80;
  let hex = "";
  for (const b of bytes) hex += b.toString(16).padStart(2, "0");
  return `anon_${hex}`;
}

function parseJSON(text: string): unknown {
  if (!text) return null;
  try {
    return JSON.parse(text);
  } catch {
    return null;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: ruby/iforevents/lib/iforevents/api_integration.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "securerandom"
require "uri"

module Iforevents
  # The first-party API integration: identify, batched track, page views, a
  # client-owned user id in X-User-Id, retries with Retry-After and typed
  # errors. Thread-safe. Mirrors IForeventsAPIIntegration of the Flutter package.
  #
  # Only +project_key+ is required. It is a public write key: it grants event
  # ingestion and nothing else. There is deliberately no project secret here.
  class APIIntegration < Integration
    USER_KEY = "iforevents_user_id"
    IDENTIFIED_KEY = "iforevents_user_identified"
    QUEUE_KEY = "iforevents_queue"
    MAX_BATCH = 500
    DEFAULT_BASE_URL = "https://api.iforevents.com"
    LIFTED_TRAITS = %w[email name phone_number].freeze

    attr_reader :project_key, :base_url, :batch_size, :flush_interval, :timeout, :max_retries, :retry_delay, :storage

    def initialize(project_key:, base_url: DEFAULT_BASE_URL, batch_size: 20, flush_interval: 10.0, timeout: 10.0, max_retries: 3, retry_delay: 1.0,
                   requeue_failed_events: true, debug: false, throw_on_error: false, on_quota_exceeded: nil, on_error: nil, storage: nil,
                   persist_queue: false, max_queue_size: 1000, user_agent: nil, flush_at_exit: true, logger: nil, **hooks)
      raise ArgumentError, "the project secret never belongs in an SDK; pass the project key only" if hooks.key?(:project_secret)
      raise ArgumentError, "APIIntegration needs a project_key" if project_key.to_s.strip.empty?

      super(name: "IForeventsAPIIntegration", **hooks)
      @project_key = project_key
      @base_url = base_url.to_s.sub(%r{/+\z}, "")
      @batch_size = batch_size.to_i.clamp(1, MAX_BATCH)
      @flush_interval = flush_interval
      @timeout = timeout
      @max_retries = [max_retries.to_i, 0].max
      @retry_delay = retry_delay
      @requeue_failed_events = requeue_failed_events
      @debug = debug
      @throw_on_error = throw_on_error
      @on_quota_exceeded = on_quota_exceeded
      @on_error = on_error
      @storage = storage || MemoryStorage.new
      @persist_queue = persist_queue
      @max_queue_size = max_queue_size
      @user_agent = user_agent || "#{SDK_NAME}/#{VERSION} ruby/#{RUBY_VERSION} (#{RbConfig::CONFIG['host_os']}; #{RbConfig::CONFIG['host_cpu']})"
      @logger = logger
      @queue = []
      @lock = Monitor.new
      @send_lock = Mutex.new
      @timer = nil
      @user_id = nil
      @initialized = false
      @identified = false
      @quota_exceeded = false
      at_exit { shutdown rescue nil } if flush_at_exit
    end

    # --- state -----------------------------------------------------------------

    def initialized?
      @initialized
    end

    def identified?
      @identified
    end

    # The id every request carries in X-User-Id: a generated "anon_..." id kept per visitor, or the custom_id of the last identify.
    def user_id
      @user_id
    end

    def queued_events
      @lock.synchronize { @queue.size }
    end

    # True after a quota_exceeded answer until the next accepted request.
    def quota_exceeded?
      @quota_exceeded
    end

    # A fresh anonymous id, unrelated to anything the server derives: anon_<uuid4 without dashes>.
    def self.anonymous_id
      "anon_#{SecureRandom.uuid.delete('-')}"
    end

    # --- Integration ------------------------------------------------------------

    def init
      super
      @lock.synchronize do
        stored = @storage.get(USER_KEY)
        if stored && !stored.empty?
          @user_id = stored
          @identified = @storage.get(IDENTIFIED_KEY) == "true"
        else
          # A fresh visitor: attribute everything to an anonymous id we own, so the
          # api never has to fingerprint the address (which merges users behind a NAT).
          set_user(self.class.anonymous_id, identified: false)
        end
        if @persist_queue && (raw = @storage.get(QUEUE_KEY))
          events = begin
            JSON.parse(raw)
          rescue JSON::ParserError
            nil
          end
          if events.is_a?(Array) && !events.empty?
            @queue = (events + @queue).last(@max_queue_size)
            schedule
          else
            @storage.remove(QUEUE_KEY)
          end
        end
        @initialized = true
      end
      debug("api integration ready base_url=#{@base_url} batch_size=#{@batch_size}")
    end

    def identify(event)
      super
      properties = event.traits.dup
      body = { "custom_id" => event.custom_id }
      LIFTED_TRAITS.each do |key|
        value = properties.delete(key) || properties.delete(key.to_sym)
        body[key] = value if value.is_a?(String) && !value.empty?
      end
      body["properties"] = properties
      # Attribute from now on, even if the profile request itself fails: the
      # api creates the profile on the first event it sees for this id.
      set_user(event.custom_id, identified: true)
      begin
        request("/v1/events/identify", body)
      rescue APIError => e
        report(e)
        raise if @throw_on_error
      end
    end

    def track(event)
      super
      queued = { "name" => event.name, "type" => event.type, "properties" => event.properties, "created_at" => iso(event.timestamp) }
      if @batch_size <= 1
        begin
          request("/v1/events/track", { "event_name" => event.name, "event_type" => event.type, "properties" => event.properties })
        rescue APIError => e
          report(e)
          raise if @throw_on_error
        end
        return
      end
      full = @lock.synchronize do
        @queue << queued
        @queue.shift(@queue.size - @max_queue_size) if @queue.size > @max_queue_size
        persist
        (@queue.size >= @batch_size).tap { |f| schedule unless f }
      end
      flush if full
    end

    def page(event)
      super
      props = event.properties.dup
      props["navigation_type"] = event.navigation_type if event.navigation_type
      props["to_route"] = event.to_route if event.to_route
      props["previous_route"] = event.previous_route if event.previous_route
      track(TrackEvent.new(name: event.name, type: "page_view", properties: props, timestamp: event.timestamp))
    end

    def reset
      super
      flush
    ensure
      # Forget the person; the next events belong to a fresh anonymous id.
      set_user(self.class.anonymous_id, identified: false)
    end

    # Sends the whole queue now, 500 events per request. Blocks until done.
    def flush
      cancel_timer
      @send_lock.synchronize do
        loop do
          events = @lock.synchronize do
            return if @queue.empty?

            @queue.shift(MAX_BATCH)
          end
          begin
            request("/v1/events/batch", { "events" => events })
            @lock.synchronize { persist }
          rescue APIError => e
            @lock.synchronize do
              if e.retryable? && @requeue_failed_events
                # Transient: keep these events at the front for the next flush.
                @queue.unshift(*events)
                schedule
              elsif !e.retryable?
                # A refused key or an exhausted quota fails the same way forever: drop everything.
                @queue.clear
              end
              persist
            end
            report(e)
            raise if @throw_on_error

            return
          end
        end
      end
    end

    def shutdown
      flush
    ensure
      cancel_timer
    end

    private

    def schedule
      return if @timer&.alive? || @queue.empty?

      interval = @flush_interval
      @timer = Thread.new do
        sleep(interval)
        @lock.synchronize { @timer = nil }
        begin
          flush
        rescue StandardError => e
          debug("timer flush failed: #{e}")
        end
      end
      @timer.abort_on_exception = false
    end

    def cancel_timer
      @lock.synchronize do
        t = @timer
        @timer = nil
        t&.kill if t && t != Thread.current
      end
    end

    def persist
      return unless @persist_queue

      if @queue.empty?
        @storage.remove(QUEUE_KEY)
      else
        @storage.set(QUEUE_KEY, JSON.generate(@queue))
      end
    end

    def set_user(id, identified:)
      @lock.synchronize do
        @user_id = id
        @identified = identified
        @storage.set(USER_KEY, id)
        @storage.set(IDENTIFIED_KEY, identified ? "true" : "false")
      end
    end

    def report(error)
      debug("request failed: #{error.message}")
      @on_error&.call(error)
    end

    def note_outcome(error)
      if error.is_a?(QuotaExceededError)
        unless @quota_exceeded
          @quota_exceeded = true
          @on_quota_exceeded&.call(error)
        end
      elsif error.nil?
        @quota_exceeded = false
      end
    end

    def debug(message)
      return unless @debug

      (@logger || $stderr).respond_to?(:info) ? @logger.info("[iforevents] #{message}") : $stderr.puts("[iforevents] #{message}")
    end

    def iso(time)
      time.respond_to?(:utc) ? time.utc.strftime("%Y-%m-%dT%H:%M:%S.%3NZ") : time.to_s
    end

    # POSTs JSON with retries; returns the decoded body (or nil).
    def request(path, body)
      payload = JSON.generate(body)
      uid = @user_id
      attempt = 0
      loop do
        error = nil
        begin
          status, text, headers = once(path, payload, uid)
          data = parse_json(text)
          debug("POST #{path} -> #{status}")
          if status.between?(200, 299)
            note_outcome(nil)
            return data
          end
          error = Iforevents.classify_response(status, data, headers["retry-after"])
        rescue SystemCallError, IOError, Timeout::Error, OpenSSL::SSL::SSLError, SocketError, Net::ProtocolError => e
          error = APIError.new(e.message)
        end
        if !error.retryable? || attempt >= @max_retries
          note_outcome(error)
          raise error
        end
        delay = error.is_a?(RateLimitedError) && error.retry_after.to_f.positive? ? error.retry_after.to_f : @retry_delay * (attempt + 1)
        debug("retrying #{path} in #{delay}s (#{attempt + 1}/#{@max_retries})")
        sleep(delay)
        attempt += 1
      end
    end

    def once(path, payload, uid)
      uri = URI.parse(@base_url + path)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      http.open_timeout = @timeout
      http.read_timeout = @timeout
      http.write_timeout = @timeout if http.respond_to?(:write_timeout=)
      req = Net::HTTP::Post.new(uri.request_uri)
      req["Content-Type"] = "application/json"
      req["X-Project-Key"] = @project_key
      req["User-Agent"] = @user_agent
      req["X-User-Id"] = uid if uid && !uid.empty?
      req.body = payload
      res = http.request(req)
      [res.code.to_i, res.body.to_s, res.to_hash.transform_values(&:first)]
    end

    def parse_json(text)
      return nil if text.nil? || text.empty?

      JSON.parse(text)
    rescue JSON::ParserError
      nil
    end
  end
end
